clifford/__init__.py

```diff
@@ -139,73 +142,70 @@ def construct_tables(
             # A_r _| B_s = <A_r B_s>_(s-r) if s-r >= 0
             lcmt_prod_mask[list_ind] = lcmt_check(grade_list_idx, grade_list_i, grade_list_j)
 
-    # Convert boolean masks to indices
-    imt_prod_mask, = imt_prod_mask.nonzero()
-    omt_prod_mask, = omt_prod_mask.nonzero()
-    lcmt_prod_mask, = lcmt_prod_mask.nonzero()
+    return indices, mult_table_vals, imt_prod_mask, omt_prod_mask, lcmt_prod_mask
 
-    return k_list, l_list, m_list, mult_table_vals, imt_prod_mask, omt_prod_mask, lcmt_prod_mask
 
+def construct_tables(
+    gradeList, linear_map_to_bitmap, bitmap_to_linear_map, signature
+) -> Tuple[sparse.COO, sparse.COO, sparse.COO, sparse.COO]:
+    # wrap the numba one
+    indices, *arrs = _numba_construct_tables(
+        gradeList, linear_map_to_bitmap, bitmap_to_linear_map, signature
+    )
+    dims = len(gradeList)
+    return tuple(
+        sparse.COO(
+            coords=indices, data=arr, shape=(dims, dims, dims),
+            prune=True
+        )
+        for arr in arrs
+    )
 
-def get_mult_function(k_list, l_list, m_list, mult_table_vals, n_dims, gradeList, product_mask=None,
+
+def get_mult_function(mt: sparse.COO, gradeList, product_mask=None,
                       grades_a=None, grades_b=None, filter_mask=None):
     '''
     Returns a function that implements the mult_table on two input multivectors
     '''
     if product_mask is not None:
-        k_list = k_list[product_mask]
-        l_list = l_list[product_mask]
-        m_list = m_list[product_mask]
-        mult_table_vals = mult_table_vals[product_mask]
+        mt = sparse.where(product_mask, mt, 0)
```

Comment:
Strictly this is a little less efficient, but it has the advantage of working with dense product masks too. At any rate, this is only a startup cost.

Reply:
How does this start up impact effect large-ish d algebras eg Cl(8)?

Reply:
Before:
```
2.16 s ± 47.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
After:
```
In [2]: %timeit clifford.Cl(8)
1.34 s ± 64.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

???

Reply:
Nice